pkg/client/client_v2.go

```diff
@@ -0,0 +1,541 @@
+	// we still want to send the responses that either failed or haven't been
+	// sent back to the agent.
+	actionResults := make(chan *proto.ActionResponse, 100)
+	go func() {
+		defer c.wg.Done()
+		for {
+			select {
+			case <-c.ctx.Done():
+				// stopped
+				return
+			default:
+			}
+
+			c.actionRoundTrip(actionResults)
+		}
+	}()
+}
+
+func (c *clientV2) actionRoundTrip(actionResults chan *proto.ActionResponse) {
+	actionsCtx, actionsCancel := context.WithCancel(c.ctx)
+	defer actionsCancel()
+	actionsClient, err := c.client.Actions(actionsCtx)
+	if err != nil {
+		c.errChan <- err
+		return
+	}
+
+	var actionsRead sync.WaitGroup
+	var actionsWrite sync.WaitGroup
+	done := make(chan bool)
+
+	// action requests
+	actionsRead.Add(1)
+	go func() {
+		defer actionsRead.Done()
+		for {
+			action, err := actionsClient.Recv()
+			if err != nil {
+				if err != io.EOF {
+					c.errChan <- err
+				}
+				close(done)
+				return
+			}
+
+			switch action.Type {
+			case proto.ActionRequest_CUSTOM:
+				c.tryPerformAction(actionResults, action)
+			case proto.ActionRequest_DIAGNOSTICS:
+				// TODO: Implement the diagnostics action.
+				// At the moment it just returns action type unknown until implemented.
+				actionResults <- &proto.ActionResponse{
+					Token:  c.token,
+					Id:     action.Id,
+					Status: proto.ActionResponse_FAILED,
+					Result: ActionTypeUnknown,
+				}
+			default:
+				actionResults <- &proto.ActionResponse{
+					Token:  c.token,
+					Id:     action.Id,
+					Status: proto.ActionResponse_FAILED,
+					Result: ActionTypeUnknown,
+				}
+			}
+		}
+	}()
+
+	// action responses
+	actionsWrite.Add(1)
+	go func() {
+		defer actionsWrite.Done()
+
+		// initial connection of stream must send the token so
+		// the Elastic Agent knows this clients token.
+		err := actionsClient.Send(&proto.ActionResponse{
+			Token:  c.token,
+			Id:     ActionResponseInitID,
+			Status: proto.ActionResponse_SUCCESS,
+			Result: []byte("{}"),
+		})
+		if err != nil {
+			c.errChan <- err
+			return
+		}
+
+		for {
+			select {
+			case <-done:
+				return
+			case res := <-actionResults:
+				err := actionsClient.Send(res)
+				if err != nil {
+					// failed to send, add back to response to try again
+					actionResults <- res
+					c.errChan <- err
+					return
+				}
+			}
+		}
+	}()
+
+	// wait for write goroutine to quit then close the client
+	actionsWrite.Wait()
+	actionsClient.CloseSend()
+
+	// then wait for the read to quit
+	actionsRead.Wait()
+}
+
+func (c *clientV2) tryPerformAction(actionResults chan *proto.ActionResponse, action *proto.ActionRequest) {
+	// find the unit
+	c.unitsLock.RLock()
+	unit := c.findUnit(action.UnitId, UnitType(action.UnitType))
+	c.unitsLock.RUnlock()
+	if unit == nil {
+		actionResults <- &proto.ActionResponse{
+			Token:  c.token,
+			Id:     action.Id,
+			Status: proto.ActionResponse_FAILED,
+			Result: ActionErrUnitNotFound,
+		}
+		return
+	}
```

Comment:
So this mean that If an agent receive an `osquery` and it receive an osquery action and the unit isn't runnning we would silently ignore the event?